UI: Update Stack component to support only gap tokens

aduth opened the pull request:
## What?

Follow-up https://github.com/WordPress/gutenberg/pull/73308#discussion_r2599189686

Updates the `Stack` component in `@wordpress/ui` to require given `gap` tokens to use one of the available design tokens, removing support for passing numeric multipliers of the base spacing unit, or any other valid flexbox `gap` value.

## Why?

Since this component is a foundational component of the design system, it should aim to build on design tokens that form the basis of expressing spacing in a systems-based manner.

This also proves to be a vast simplification of the component, which previously had to do some complex internal logic to juggle different forms of `gap` values.

## Testing Instructions

Repeat Testing Instructions from #73308

aduth committed:
UI: Update Stack component to support only gap tokens

aduth commented on packages/ui/src/stack/stories/index.story.tsx:
```diff
@@ -121,7 +109,7 @@ export const Nested: Story = {
 				</Stack>
 				<DemoBox variant="lg" />
-				<Stack gap={ 0 } direction="column">
+				<Stack gap="2xs" direction="column">
```
I wondered if we'd need some sort of "none" token, or explicitly allow a `0` value, but then I recalled that the default would be to have no gap when `undefined`, which is what we want (i.e. `gap={ 0 }` is the same as leaving out `gap` altogether). This is good affirmation of choosing zero spacing as the default gap behavior.

jameskoster replied:
Do you think we should include a story to at least demonstrate that zero gap is possible?

aduth replied:
I'll restore this to effectively on gap as an example (updated in 762883295fd).

I could maybe see a case that our "Default" story should actually reflect defaults and not specify the gap, but I feel like it's more practically useful to demonstrate the component with some gap. Open to revisiting.

tyxla reviewed (approved):
That's a nice simplification 👍 

Tests as promised ✅

aduth committed:
Remove gap from nested Stack story example

https://github.com/WordPress/gutenberg/pull/73852#discussion_r2603902080

Include a Storybook example that has no gap, and ensure that the component can be used without passing gap
